skfem/io/meshio.py

```diff
@@ -130,3 +132,23 @@ def find_tagname(tag):
 
 def from_file(filename):
     return from_meshio(meshio.read(filename))
+
+
+def to_file(mesh, filename, point_data=None, cell_data=None, **kwargs):
+
+    meshio_type = TYPE_MESH_MAPPING[type(mesh)]
+
+    if point_data is not None:
+        if not isinstance(point_data, dict):
+            raise ValueError("point_data should be "
+                             "a dictionary of ndarrays.")
+
+    if cell_data is not None:
+        if not isinstance(cell_data, dict):
+            raise ValueError("cell_data should be "
+                             "a dictionary of ndarrays.")
+        cell_data = {meshio_type: cell_data}
+
+    cells = {meshio_type: mesh.t.T}
```

Comment:
The `meshio.Mesh.cells` attribute has changed in meshio 4.  It's now `List[CellBlock]` where 
```python
CellBlock = collections.namedtuple("CellBlock", ["type", "data"])
```

https://github.com/nschloe/meshio/blob/add5d452f853cca3f2cb4a9151b297dda03442ce/meshio/_mesh.py#L5

and `type: str` ('line', 'triangle', …) and `data: ndarray` (`ndim == 2`, dtype some kind of int, with one cell per column).

This old type (`Dict[str, ndarray]`) is still tolerated and converted in the initializer but deprecated in comments.

https://github.com/nschloe/meshio/blob/add5d452f853cca3f2cb4a9151b297dda03442ce/meshio/_mesh.py#L22-L35

I think the up-to-date thing here would be
```python
cells = [(meshio_type, mesh.t.T)]
```

Reply:
The situation in meshio 4 with `meshio.Mesh.cell_data` is similar, except that it's not converted in the constructor; anything truthy is just accepted.

https://github.com/nschloe/meshio/blob/add5d452f853cca3f2cb4a9151b297dda03442ce/meshio/_mesh.py#L37

Assuming
```python
cells = [(meshio.type, mesh.t.T)]
```
that means something like
```python
cell_data = {k: [(meshio_type, v)] for k, v in cell_data.items()}
```

So far there are no examples of `cell_data` (or `**kwargs`) in scikit-fem/docs/examples, only point_data.

Reply:
I'll add a test for `cell_data` at least.

Reply:
We _could_ also start saving `skfem.Mesh.boundaries` as `meshio.Mesh.cell_sets`, although it's not yet used by `meshio.write` for many formats. Anyway, that needn't hold up this PR.

Reply:
I'll see if I have time to try it out. So I should be able to save and load named boundaries?

Reply:
Oops, sorry, I meant `skfem.Mesh.subdomains`, not `boundaries`; the latter are more complicated again as they involve facets rather than cells.  meshio doesn't support facets. nschloe/meshio#729

But even with `subdomains` it depends on one's interpretation of ‘should’.  I think that the intent of the newish attribute `meshio.Mesh.cell_sets` is very similar to `skfem.Mesh.subdomains`; see 

https://github.com/kinnala/scikit-fem/blob/ada7b469bd7ce5eab58677cf8a073d67e454b41c/skfem/io/meshio.py#L79-L82

for how we can read `skfem.Mesh.subdomains` from `meshio.Mesh.cell_sets_dict`, which is a property that restructures `meshio.Mesh.cell_sets`.  We should be able to invert this with something like
```python
cell_sets = {k: [(meshio_type, v)] for k, v in mesh.boundaries.items()}
```
however, as [mentioned](https://github.com/kinnala/scikit-fem/pull/350#discussion_r393537114), I  don't think `meshio.write` actually writes the `meshio.Mesh.cell_sets` to very many formats, maybe only Abaqus, which I haven't tried.  Not many mesh file formats could support the functionality of cell_sets; MSH 4.1 could but meshio doesn't really write that format.

But hopefully in the future meshio will write `cell_sets` to more formats (maybe XDMF, Exodus, …?) and that will give more options for serialization #158.

Without support from `meshio.write`, there might still be use in defining `cell_sets` in terms of `boundaries` if something else is to be done with the `meshio.Mesh`.  There might be external libraries that recognize and process `meshio.Mesh` (smoothers? hex-splitters? I don't know).  This would necessitate factoring a function `skfem.io.meshio.to_meshio` out of `to_file` in the same way that `from_meshio` was factored out of `from_file`.

Sorry, these are vague optimistic speculations on the future which aren't really related to this PR.